vernac/comDefinition.ml

```diff
@@ -48,57 +48,59 @@ let warn_implicits_in_term =
-        (* Second, we need to prepare the universes, including minimization. Not fun *)
-        let evd,_ = Evarutil.nf_evars_and_universes evd in
-        let vars = EConstr.universes_of_constr env evd body in
-        let evd = Evd.restrict_universe_context evd vars in
-        let uctx = Evd.check_univ_decl ~poly evd decl in
-        (* Third, we substitute all the evars to store in the kernel. *)
-        let body = EConstr.to_constr evd body in
-        (* Fourth, store *)
-        evd, imps1@(Impargs.lift_implicits nb_args imps2),
-          definition_entry ~univs:uctx body
-    | Some ctyp ->
-        (* 1 *)
-        let evd, (ty, impsty) = interp_type_evars_impls ~impls env_bl evd ctyp in
-        let env_bl = EConstr.push_rel_context ctx env in
-        let evd, (c, imps2) = interp_casted_constr_evars_impls ~impls env_bl evd c ty in
-        let body = EConstr.(it_mkLambda_or_LetIn c ctx) in
-        let typ = EConstr.(it_mkProd_or_LetIn ty ctx) in
-        (* 2 *)
-        let beq b1 b2 = if b1 then b2 else not b2 in
-        let impl_eq (x,y,z) (x',y',z') = beq x x' && beq y y' && beq z z' in
-        (* Check that all implicit arguments inferable from the term
-           are inferable from the type *)
-        let chk (key,va) =
-          impl_eq (List.assoc_f Pervasives.(=) key impsty) va (* FIXME *)
-        in
-        if not (try List.for_all chk imps2 with Not_found -> false)
-        then warn_implicits_in_term ();
-        (* 3 *)
-        let evd, _ = Evarutil.nf_evars_and_universes evd in
-        let bodyvars = EConstr.universes_of_constr env evd body in
-        let tyvars = EConstr.universes_of_constr env evd ty in
-        let vars = Univ.LSet.union bodyvars tyvars in
-        let evd = Evd.restrict_universe_context evd vars in
-        let uctx = Evd.check_univ_decl ~poly evd decl in
-        (* 4 *)
-        let body,typ = EConstr.(to_constr evd body, to_constr evd typ) in
-        evd, imps1@(Impargs.lift_implicits nb_args impsty),
-          definition_entry ~types:typ ~univs:uctx body
+  (* Build the type *)
+  let evd, tyopt =
+    Option.cata (fun ctyp ->
+        let evd, ty = interp_type_evars_impls ~impls env_bl evd ctyp in
+        evd, Some ty)
+      (evd, None) ctypopt
+  in
+  (* Build the body, and merge implicits from parameters and from type/body *)
+  let evd, c, imps, tyopt =
+    match tyopt with
+    | None ->
+      let evd, (c, impsbody) = interp_constr_evars_impls ~impls env_bl evd c in
+      evd, c, imps1@Impargs.lift_implicits (Context.Rel.nhyps ctx) impsbody, None
+    | Some (ty, impsty) ->
+      let evd, (c, impsbody) = interp_casted_constr_evars_impls ~impls env_bl evd c ty in
+      check_imps ~impsty ~impsbody;
+      evd, c, imps1@Impargs.lift_implicits (Context.Rel.nhyps ctx) impsty, Some ty
+  in
+  (* universe minimization *)
+  let evd = Evd.nf_constraints evd in
+  (* Unless program mode the body and type are ready, substitute evars *)
```

Comment:
What does "unless program mode" mean here? Maybe you want to expand the comment?